File: scripts/seed-super-user.mjs
#!/usr/bin/env node

import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";
import mongoose from "mongoose";
import bcrypt from "bcryptjs";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

loadEnvFile(path.resolve(__dirname, "..", ".env.local"));

const {
  MONGODB_URI,
  DB_CSW_NAME,
  MONGODB_DB,
  SUPER_USER_MAIL,
  SUPER_USER_PASSWORD,
} = process.env;

if (!MONGODB_URI) {
  console.error("Missing MONGODB_URI environment variable.");
  process.exit(1);
}

if (!SUPER_USER_MAIL || !SUPER_USER_PASSWORD) {
  console.error("SUPER_USER_MAIL and SUPER_USER_PASSWORD must be set in your environment.");
  process.exit(1);
}

async function main() {
  const connectionOptions = {
    bufferCommands: false,
    ...((DB_CSW_NAME || MONGODB_DB) ? { dbName: DB_CSW_NAME ?? MONGODB_DB } : {}),
  };

  await mongoose.connect(MONGODB_URI, connectionOptions);

  const userSchema = new mongoose.Schema(
    {
      email: { type: String, required: true, unique: true, lowercase: true, trim: true, index: true },
      passwordHash: { type: String, required: true },
      name: { type: String, trim: true },
      roles: { type: [String], default: [] },
      refreshTokenVersion: { type: Number, default: 0 },
      resetTokenVersion: { type: Number, default: 0 },
    },
    { timestamps: true },
  );

  const UserModel = mongoose.models.User ?? mongoose.model("User", userSchema);

  const email = SUPER_USER_MAIL.toLowerCase();
  const existingUser = await UserModel.findOne({ email }).exec();

  const passwordHash = await bcrypt.hash(SUPER_USER_PASSWORD, 12);

  if (existingUser) {
    const shouldUpdate = !(await bcrypt.compare(SUPER_USER_PASSWORD, existingUser.passwordHash));
    if (shouldUpdate) {
      existingUser.passwordHash = passwordHash;
      existingUser.roles = Array.from(new Set([...(existingUser.roles ?? []), "super-admin"]));
      await existingUser.save();
      console.log(`Updated existing super user password and roles for ${email}`);
    } else {
      if (!existingUser.roles?.includes("super-admin")) {
        existingUser.roles = [...(existingUser.roles ?? []), "super-admin"];
        await existingUser.save();
        console.log(`Added super-admin role to existing user ${email}`);
      } else {
        console.log(`Super user ${email} already exists. No changes needed.`);
      }
    }
  } else {
    await UserModel.create({
      email,
      passwordHash,
      roles: ["super-admin"],
      name: "Super Admin",
    });
    console.log(`Created super user ${email}`);
  }

  await mongoose.disconnect();
}

main()
  .then(() => {
    console.log("Seeding complete.");
    process.exit(0);
  })
  .catch(async (error) => {
    console.error("Failed to seed super user:", error);
    try {
      await mongoose.disconnect();
    } catch {}
    process.exit(1);
  });

function loadEnvFile(filePath) {
  if (!fs.existsSync(filePath)) {
    return;
  }

  const contents = fs.readFileSync(filePath, "utf-8");
  for (const line of contents.split(/\r?\n/)) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) {
      continue;
    }

    const equalsIndex = trimmed.indexOf("=");
    if (equalsIndex === -1) {
      continue;
    }

    const key = trimmed.slice(0, equalsIndex).trim();
    let value = trimmed.slice(equalsIndex + 1).trim();

    if ((value.startsWith("\"") && value.endsWith("\"")) || (value.startsWith("'") && value.endsWith("'"))) {
      value = value.slice(1, -1);
    }

    if (!(key in process.env)) {
      process.env[key] = value;
    }
  }
}

File: scripts/seed_super_user.py
"""Seed or update a CSW super admin user in MongoDB."""

from __future__ import annotations

import os
import sys
from datetime import UTC, datetime
from typing import Any, Dict, Optional

import bcrypt
from pymongo import MongoClient
from pymongo.collection import Collection
from pymongo.database import Database
from pymongo.errors import ConfigurationError, PyMongoError


REQUIRED_ENV_KEYS = ("MONGODB_URI", "SUPER_USER_MAIL", "SUPER_USER_PASSWORD")


def _get_required_env(key: str) -> str:
    value = os.getenv(key)
    if not value:
        raise SystemExit(f"Missing {key} environment variable.")
    return value


def _resolve_database(client: MongoClient) -> Database:
    db_name = os.getenv("DB_CSW_NAME") or os.getenv("MONGODB_DB")

    if db_name:
        return client.get_database(db_name)

    try:
        return client.get_default_database()
    except ConfigurationError as exc:  # pragma: no cover - informative failure path
        raise SystemExit(
            "MongoDB database name not provided. Set DB_CSW_NAME or include it in MONGODB_URI."
        ) from exc


def _ensure_super_admin(users: Collection, email: str, password: str) -> None:
    now = datetime.now(UTC)
    hashed_password = bcrypt.hashpw(password.encode("utf-8"), bcrypt.gensalt(rounds=12)).decode("utf-8")

    existing: Optional[Dict[str, Any]] = users.find_one({"email": email})
    if existing:
        updates: Dict[str, Any] = {}

        stored_hash = existing.get("passwordHash")
        if not stored_hash or not bcrypt.checkpw(password.encode("utf-8"), stored_hash.encode("utf-8")):
            updates["passwordHash"] = hashed_password

        roles = set(existing.get("roles", []))
        if "super-admin" not in roles:
            roles.add("super-admin")
            updates["roles"] = sorted(roles)

        if updates:
            updates["updatedAt"] = now
            users.update_one({"_id": existing["_id"]}, {"$set": updates})
            print(f"Updated super admin user {email}")
        else:
            print(f"Super admin user {email} already up to date")
        return

    document = {
        "email": email,
        "passwordHash": hashed_password,
        "roles": ["super-admin"],
        "name": "Super Admin",
        "refreshTokenVersion": 0,
        "resetTokenVersion": 0,
        "createdAt": now,
        "updatedAt": now,
    }
    users.insert_one(document)
    print(f"Created super admin user {email}")


def main() -> None:
    for key in REQUIRED_ENV_KEYS:
        _get_required_env(key)

    uri = os.environ["MONGODB_URI"]
    email = os.environ["SUPER_USER_MAIL"].strip().lower()
    password = os.environ["SUPER_USER_PASSWORD"]

    client = MongoClient(uri, tz_aware=True)
    try:
        database = _resolve_database(client)
        users_collection = database.get_collection("users")
        _ensure_super_admin(users_collection, email, password)
    except PyMongoError as exc:
        raise SystemExit(f"Failed to seed super admin: {exc}") from exc
    finally:
        client.close()


if __name__ == "__main__":
    try:
        main()
    except SystemExit as exc:
        print(exc, file=sys.stderr)
        raise
